### PR TITLE
Refine validation error handling

### PR DESCRIPTION
**Pull request by sedroche**
Validation was setting the failed validation metric on err.

- I have refined the flow to only set validation metric on validation failure
- One new test for setting the failed alert metric correctly

### CODE REVIEW

**Comment by sedroche** [comment]
cc @dofinn

File: pkg/validation/validation.go [hunk 1]
```diff
@@ -25,56 +25,94 @@ func NewBuilder() ValidationBuilder {
...
 type Validator interface {
-	IsValidUpgradeConfig(uC *upgradev1alpha1.UpgradeConfig, cV *configv1.ClusterVersion, logger logr.Logger) (bool, error)
+	IsValidUpgradeConfig(uC *upgradev1alpha1.UpgradeConfig, cV *configv1.ClusterVersion, logger logr.Logger) (ValidatorResult, error)
 }
 
 type validator struct{}
 
-func (v *validator) IsValidUpgradeConfig(uC *upgradev1alpha1.UpgradeConfig, cV *configv1.ClusterVersion, logger logr.Logger) (bool, error) {
+type ValidatorResult struct {
+	IsRollBack bool
```
Comment (dofinn) [inline pkg/validation/validation.go:34]: Where is this used? What is its purpose?

**Comment by wanghaoran1988** [comment]
If first time the validation failed for some reason, and then succeed next time, will the metrics be updated? if not, do we have alert use this metrics? and will the alert be cleared?

**Comment by sedroche** [comment]
> If first time the validation failed for some reason, and then succeed next time, will the metrics be updated? if not, do we have alert use this metrics? and will the alert be cleared?

I believe the prometheus rules only fire after 5 mins:
https://github.com/openshift/managed-cluster-config/blob/88e04b670c3b3bd05416143ef2e1271720bab9a2/deploy/sre-prometheus/100-managed-upgrade-operator.PrometheusRule.yaml#L16

we reset the metric after its validated:
https://github.com/openshift/managed-upgrade-operator/blob/86c096884438cec038b431017bec372ce614336b/pkg/controller/upgradeconfig/upgradeconfig_controller.go#L166

File: pkg/validation/validation.go (hunk 1 above)
Comment (sedroche) [reply pkg/validation/validation.go:34]: Good spot.
There was a scenario of `false, nil` for roll back which I tried to capture.
Ok to remove iyo?

File: pkg/controller/upgradeconfig/upgradeconfig_controller_test.go [hunk 2]
```diff
@@ -237,6 +238,20 @@ var _ = Describe("UpgradeConfigController", func() {
 						},
 					}
 				})
+				Context("When the upgradeconfig validation fails", func() {
+					It("should set set the validation alert metric", func() {
```
Comment (bmeng) [inline pkg/controller/upgradeconfig/upgradeconfig_controller_test.go:242]: dupe "set"

File: pkg/validation/validation.go (hunk 1 above)
Comment (bmeng) [reply pkg/validation/validation.go:34]: Will we support rollback? Do we need to separate this condition out?

Comment (sedroche) [reply pkg/validation/validation.go:34]: @bmeng @dofinn I'm going to remove this

File: pkg/controller/upgradeconfig/upgradeconfig_controller_test.go (hunk 2 above)
Comment (sedroche) [reply pkg/controller/upgradeconfig/upgradeconfig_controller_test.go:242]: Thank you

**Pull request merged**